test/ranges.coffee

```diff
@@ -116,3 +116,99 @@ test "#1012 slices with arguments object", ->
 
 test "#1409: creating large ranges outside of a function body", ->
   CoffeeScript.eval '[0..100]'
+
+test "#2047: Infinite loop possible when `for` loop with `range` uses variables", ->
+  up = 1
+  down = -1
+  a = 1
+  b = 5
+
+  testRange = (arg) ->
+    [from, to, step, expectedResult] = arg
+    r = (x for x in [from..to] by step)
+    arrayEq r, expectedResult
+
+  testData = [
+    [1, 5, 1, [1..5]]
+    [1, 5, -1, [1]]
+    [1, 5, up, [1..5]]
+    [1, 5, down, [1]]
+
+    [a, 5, 1, [1..5]]
+    [a, 5, -1, [1]]
+    [a, 5, up, [1..5]]
+    [a, 5, down, [1]]
+
+    [1, b, 1, [1..5]]
+    [1, b, -1, [1]]
+    [1, b, up, [1..5]]
+    [1, b, down, [1]]
+
+    [a, b, 1, [1..5]]
+    [a, b, -1, [1]]
+    [a, b, up, [1..5]]
+    [a, b, down, [1]]
+
+    [5, 1, 1, [5]]
+    [5, 1, -1, [5..1]]
+    [5, 1, up, [5]]
+    [5, 1, down,  [5..1]]
+
+    [5, a, 1, [5]]
+    [5, a, -1, [5..1]]
+    [5, a, up, [5]]
+    [5, a, down, [5..1]]
+
+    [b, 1, 1, [5]]
+    [b, 1, -1, [5..1]]
+    [b, 1, up, [5]]
+    [b, 1, down, [5..1]]
+
+    [b, a, 1, [5]]
+    [b, a, -1, [5..1]]
+    [b, a, up, [5]]
+    [b, a, down, [5..1]]
+  ]
+
+  testRange d for d in testData
+
+test "#2047: from, to and step as variables", ->
+  up = 1
+  down = -1
+  a = 1
+  b = 5
+
+  r = (x for x in [a..b] by up)
+  arrayEq r, [1..5]
+
+  r = (x for x in [a..b] by down)
+  arrayEq r, [1]
+
+  r = (x for x in [b..a] by up)
+  arrayEq r, [5]
+
+  r = (x for x in [b..a] by down)
+  arrayEq r, [5..1]
+
+  a = 1
+  b = -1
+  step = 0
+  r = (x for x in [b..a] by step)
+  arrayEq r, []
+
+  a = "aaa"
+  b = "bbb"
+  step = "3"
+  r = (x for x in [b..a] by step)
+  arrayEq r, ['bbb']
+
+  step = "ccc"
+  r = (x for x in [b..a] by step)
+  arrayEq r, []
+
+  a = 2
+  b = 1
+  step = "0.5000000"
+  r = (x for x in [b..a] by step)
+  arrayEq r, [1, 1.5, 2]
```

Comment:
Whoa, we want `by` to convert strings to floats?

Reply:
Actually no.  While working on this PR I had a case with an infinite loop if `step` was a string.
The idea I had was to cast `step` into the number to avoid infinite loops by accidents, e.g. in case user makes a mistake. 
But, now I see the error in my case was due to something else and this is not needed.
I'll correct this.